app/services/combined_audio_derivative_creator.rb

```diff
@@ -0,0 +1,163 @@
+require 'tempfile'
+#
+# Generate the combined audio derivatives for a
+# given oral history work.
+#
+# some_work = Work.find_by_friendlier_id(friendlier_id)
+# CombinedAudioDerivativeCreator.new(some_work).generate
+#
+#
+# Sample output:
+#     {
+#      :combined_audio_mp3_data=>
+#          "/var/folders/n7/t4zt2w751bj_6pbnf95sgpnrmz8lz8/T/output20200228-61637-ec6kjz.mp3",
+#      :combined_audio_webm_data=>
+#          "/var/folders/n7/t4zt2w751bj_6pbnf95sgpnrmz8lz8/T/output20200228-61637-1btjscz.webm",
+#      :fingerprint=>"4998d366f5edb6222db1181c7b153e21",
+#      :component_metadata=>
+#        {:durations=>
+#          [
+#            "00:30:30.29",
+#            "00:29:13.32",
+#            "00:33:03.26",
+#            "00:24:28.94"
+#          ]
+#        }
+#     }
+#
+#
+class CombinedAudioDerivativeCreator
+
+  attr_reader :work
+
+  def initialize(work)
+    @cmd = TTY::Command.new(printer: :null)
+    @work = work
+    @components = download_components
+  end
+
+  def generate
+    # Extract duration metadata for each component:
+    durations = @components.map do |f|
+      duration_of_audio_file_in_seconds(f.path)
+    end
+
+    output_paths = {}
+
+    # Create the two output files:
+    ['mp3', 'webm'].each do |format|
+      output_paths[format] = output_file(format)
+      ffmpeg_args = args_for_ffmpeg(output_paths[format])
+      log "Creating #{format} using: #{ffmpeg_args.join(" ")}"
+      @cmd.run(*ffmpeg_args, binmode: true)
+    end
+
+    # Get rid of the downloaded originals:
+    @components.map!(&:unlink)
+
+    # Return the metadata, including paths to the two output files:
+    {
+      combined_audio_mp3_data:   output_paths['mp3'],
+      combined_audio_webm_data:  output_paths['webm'],
+      fingerprint: fingerprint,
+      component_metadata: {durations: durations}
+    }
+  end
+
+  def output_file(format)
+    Tempfile.new(['output', ".#{format}"], :encoding => 'binary').path
+  end
+
+  # Use ffprobe to determine the length of an audio file.
+  def duration_of_audio_file_in_seconds(path)
+    options = ['ffprobe', '-v', 'error',
+      '-show_entries', 'format=duration', '-of',
+      'default=noprint_wrappers=1:nokey=1'
+    ] + [ path ]
+    @cmd.run(*options).out.strip
+  end
+
+  def download_components
+    result = []
+    audio_member_files.each do |original_file|
+      log "Downloading #{original_file.metadata['filename']}"
+      new_temp_file = Tempfile.new(['temp_', original_file.metadata['filename'].downcase], :encoding => 'binary')
+      original_file.open(rewindable:false) do |input_audio_io|
+        new_temp_file.write input_audio_io.read until input_audio_io.eof?
+      end
+      result << new_temp_file
+    end
+    result
```

Comment:
As this method is called in the constructor it means just constructing one of these objects will trigger download of all components. That seems like a bad pattern to me, the "expensive" thing shouldn't happen in the constructor, it should only happen when you call some "action" method. I think that will be more intuitively expected behavior. 

While you could make this method lazy, I think it may make sense to keep the list of actual Asset models that will be used internally, and then only download them into files later. This will make it easier to get metadata from them as needed (like to use as keys above in duration!), to have them all there. We may also have duration already stored in metadata as an asset, so you don't need to use a shell out to ffmpeg to look it up?  (But if we don't, that's cool).

Reply:
Removed the method from the constructor, and memoized the `components` method (not the `audio_member_files` method, since that is only called once, from the `components` method).